### PR TITLE
Detect command interface 

### PR DESCRIPTION
**Pull request by gastonrod07**
- Adds a detect command interface
- Also modifies some logic in web.py

### CODE REVIEW

File: luminoth/utils/predict.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                    config.train.job_dir, config.train.run_name))
+                if not ckpt or not ckpt.all_model_checkpoint_paths:
+                    raise ValueError('Could not find checkpoint in {}.'.format(
+                        config.train.job_dir
+                    ))
+                ckpt = ckpt.all_model_checkpoint_paths[-1]
+                ckpt_dir = os.path.join('.', ckpt)
+                saver = tf.train.Saver(sharded=True, allow_empty=True)
+                saver.restore(session, ckpt_dir)
+            else:
```
Comment (vierja) [inline luminoth/utils/predict.py]: Add a comment that predictions without checkpoints are only used for testing.

File: luminoth/utils/predict.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    else:
+        objects_labels = objects_labels.tolist()
+
+    return {
+        'objects': objects.tolist(),
+        'objects_labels': objects_labels,
+        'objects_labels_prob': objects_labels_prob.tolist(),
+        'inference_time': end_time - start_time,
+        'scale_factor': scale_factor,
+        'image_tensor': image_tensor,
```
Comment (vierja) [inline luminoth/utils/predict.py]: Make `image_tensor`, `output` and `session` optional return values.

Comment (vierja) [reply luminoth/utils/predict.py]: Also, rename `output` key to `prediction_dict`.

**Comment by dekked** [comment]
I think the command should be called `predict` and not `detect`.

This way, we can use it for whatever task (classification, object detection, masking, pose estimation, etc.) and this would only be differentiated by choosing a different  `--model` parameter.

File: luminoth/detect.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import easydict
+import tensorflow as tf
+import os
+import yaml
+
+from luminoth.utils.predict import get_prediction
+from PIL import Image, ImageDraw
+
+
+@click.command(help='Detects objects in a image.')
```
Comment (dekked) [inline luminoth/detect.py]: `Obtain a model's predictions in a given image.`

File: luminoth/detect.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+@click.argument('image-path')
+@click.option('--config-file')
+@click.option('--model-name', default='fasterrcnn')
+def detect(image_path, config_file, model_name):
```
Comment (dekked) [inline luminoth/detect.py]: This should primarily output a JSON file with the predictions. Think that somebody will want to use this programmatically, and giving this person the images will not do any good (unless for debugging).

There should be an argument (disabled by default) that lets you also output the image with the predictions drawn on top, but this should be optional.

Comment (dekked) [reply luminoth/detect.py]: Optionally, this could also be a separate command that takes an input image, and predictions (in json format) and draws whatever it has to draw over the image.

To discuss.

File: luminoth/detect.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import yaml
+
+from luminoth.utils.predict import get_prediction
+from PIL import Image, ImageDraw
+
+
+@click.command(help='Detects objects in a image.')
+@click.argument('image-path')
+@click.option('--config-file')
+@click.option('--model-name', default='fasterrcnn')
```
Comment (dekked) [inline luminoth/detect.py]: This should also work with image classification. /cc @vierja

File: luminoth/utils/predict.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    return image_array, upscale * downscale
+
+
+def get_prediction(model_name, image, config_file, session=None,
+                   prediction_dict=None, image_tensor=None,
+                   return_tf_vars=False):
+    """
+    Gets the prediction given by the model `model_name` of the image `image`.
+    If it exists a checkpoint loads it and also loads the name of the
+    classes from the dataset directory.
```
Comment (dekked) [inline luminoth/utils/predict.py]: Make more explicit. "Returns a dictionary with...".

Fix grammar & more:
```
If a checkpoint exists in the job's directory, load it.
The names of the classes will be obtained from the dataset directory.
```

File: luminoth/utils/predict.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+    image = image.resize((new_width, new_height))
+    image_array = np.array(image)[:, :, :3]  # TODO Read RGB
+    image_array = np.expand_dims(image_array, axis=0)
+    return image_array, upscale * downscale
+
+
+def get_prediction(model_name, image, config_file, session=None,
+                   prediction_dict=None, image_tensor=None,
+                   return_tf_vars=False):
```
Comment (vierja) [inline luminoth/utils/predict.py]: Add docstring with explanation, especially for the `return_tf_vars`.

**Comment by vierja** [comment]
@dekked @gastonrod07 I agree with Alan's comments.

File: luminoth/detect.py (hunk 5 above)
Comment (vierja) [reply luminoth/detect.py]: `model-name` should be `model-type` but I'm going to remove it as a CLI option in #91.
